doc/designs/quic-design/quic-api.md

```diff
@@ -818,6 +818,219 @@ reason string in bytes.
+connection SSL objects efficiently for new stream and stream readability events.
+This is not yet urgent but will be more relevant for concurrent server
+applications.
+
+#### `SSL_get0_connection`
+
+```c
+/*
+ * Get the SSL object representing the connection associated with this object.
+ *
+ * If the SSL object represents a non-QUIC method or a QUIC connection, this
+ * returns the same object passed.
+ *
+ * If the SSL object represents a QUIC stream returns the QUIC connection
+ * object.
+ */
+SSL *SSL_get0_connection(SSL *ssl);
+```
+
+#### `SSL_is_connection`
+
+```c
+/*
+ * Returns 1 if the object represents a connection. This always returns 1 for
+ * non-QUIC methods, but returns 0 for SSL objects for QUIC streams which are
+ * not also the QUIC connection object.
+ *
+ * This is exactly equivalent to (SSL_get0_connection(ssl) == ssl).
+ */
+int SSL_is_connection(SSL *ssl);
+```
+
+#### `SSL_get_stream_type`
+
+```c
+/*
+ * If the object represents a stream, returns a SSL_STREAM_TYPE value
+ * designating whether the stream can be used for transmission, reception,
+ * or both.
+ *
+ * This always returns SSL_STREAM_TYPE_BIDI for non-QUIC methods.
+ *
+ * It returns SSL_STREAM_TYPE_NONE for a QUIC connection object if it
+ * does not have a default stream.
+ */
+#define SSL_STREAM_TYPE_NONE    0
+#define SSL_STREAM_TYPE_READ    1
+#define SSL_STREAM_TYPE_WRITE   2
+#define SSL_STREAM_TYPE_BIDI    (SSL_STREAM_TYPE_READ | SSL_STREAM_TYPE_WRITE)
+__owur int SSL_get_stream_type(SSL *ssl);
+```
+
+#### `SSL_get_stream_id`
+
+```c
+/*
+ * QUIC: Returns the unique stream ID for the stream, an integer in range [0,
+ * 2**62-1]. If called on a QUIC connection, returns the unique stream ID for
+ * the default stream if there is one, and otherwise returns 0.
+ *
+ * TLS, DTLS: Returns 0.
+ */
+__owur uint64_t SSL_get_stream_id(SSL *ssl);
+```
+
+#### `SSL_new_stream`
+
+```c
+/*
+ * Create a new SSL object representing a single additional stream.
+ *
+ * There is no need to call SSL_connect on the resulting object, and
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Creates a new stream. Must be called only on a QUIC connection SSL object.
+ *   Can be used on client or server. If the SSL_FLAG_UNI flag is set, the
+ *   created stream is unidirectional, otherwise it is bidirectional.
+ *
+ *   To be MSMT-safe.
+ *
+ * For TLS and DTLS SSL objects:
+ *   Always fails.
+ */
+#define SSL_FLAG_UNI    1
+
+SSL *SSL_new_stream(SSL *ssl, uint64_t flags);
+```
+
+#### `SSL_accept_stream`
+
+```c
+/*
+ * Create a new SSL object representing an additional stream which has created
+ * by the peer.
+ *
+ * There is no need to call SSL_accept on the resulting object, and
+ * any such call is a no-op.
+ *
+ * For QUIC:
+ *   Must be called only on a QUIC connection SSL object. Fails if called on a
+ *   stream object. Checks if a new stream has been created by the peer. If it
+ *   has, creates a new SSL object to represent it and returns it. Otherwise,
```

Comment:
Is it worth defining what happens when there are multiple new streams pending?  I.e. it picks the oldest or one randomly or whatever?